src/Core/AdminConsole/OrganizationFeatures/Policies/PolicyValidators/TwoFactorAuthenticationPolicyValidator.cs

```diff
@@ -55,33 +66,66 @@ private async Task RemoveNonCompliantUsersAsync(Guid organizationId)
         var savingUserId = _currentContext.UserId;
 
         var orgUsers = await _organizationUserRepository.GetManyDetailsByOrganizationAsync(organizationId);
-        var organizationUsersTwoFactorEnabled = await _twoFactorIsEnabledQuery.TwoFactorIsEnabledAsync(orgUsers);
-        var removableOrgUsers = orgUsers.Where(ou =>
-            ou.Status != OrganizationUserStatusType.Invited && ou.Status != OrganizationUserStatusType.Revoked &&
-            ou.Type != OrganizationUserType.Owner && ou.Type != OrganizationUserType.Admin &&
-            ou.UserId != savingUserId);
-
-        // Reorder by HasMasterPassword to prioritize checking users without a master if they have 2FA enabled
-        foreach (var orgUser in removableOrgUsers.OrderBy(ou => ou.HasMasterPassword))
+
+        // this can get moved into the private method after AccountDeprovisiong flag check is removed.
+        var organizationUsersTwoFactorEnabled = (await _twoFactorIsEnabledQuery.TwoFactorIsEnabledAsync(orgUsers)).ToList();
+
+        var revocableOrgUsers = orgUsers.Where(ou =>
+                ou.Status != OrganizationUserStatusType.Invited && ou.Status != OrganizationUserStatusType.Revoked &&
+                ou.Type != OrganizationUserType.Owner && ou.Type != OrganizationUserType.Admin &&
+                ou.UserId != savingUserId)
+            .ToList();
+
+        if (_featureService.IsEnabled(FeatureFlagKeys.AccountDeprovisioning))
+        {
+            if (DoMembersNotHaveAPasswordWithTwoFactorAuthenticationEnabled(revocableOrgUsers, organizationUsersTwoFactorEnabled))
+            {
+                throw new BadRequestException(NonCompliantMembersWillLoseAccess);
+            }
+
+            var result = await _revokeNonCompliantOrganizationUserCommand.RevokeNonCompliantOrganizationUsersAsync(
+                new RevokeOrganizationUsersRequest(organizationId, revocableOrgUsers,
+                    new StandardUser(savingUserId ?? Guid.Empty,
+                        await _currentContext.OrganizationOwner(organizationId))));
```

Comment:
So here's a fun one. `currentContext.OrganizationOwner` implicitly includes providers. This is... _somewhat_ known when dealing with `currentContext`. But I recommend avoiding this ambiguity in your new object - e.g. adding an xmldoc noting that `IsOwner` can include providers in this context.

Reply:
I'll do you one better. I'll change it to `IsOwnerOrProvider`